File: features/CommandsAndShitterList.js
import utils from "../utils.js"
import Settings from "../config.js"
import { startMovingGui } from "./Gui.js"

function tyfr() {
    ChatLib.command("pc tyfr ( ﾟ◡ﾟ)/")
    setTimeout(() => {
        ChatLib.command("p leave")
    }, 500);
}


//Get shitters
let shitters = []
try {
    shitters = FileLib.read("./config/ChatTriggers/modules/shaweelAddons/shitters.json")
    shitters = JSON.parse(shitters)
} catch (err) {
    utils.errorLog(err)
    utils.errorLog("Failed to retrieve shitters. Your shitter list is treated as empty.")
}

//Kick shitters
register("chat", (player, dungeonClass, level) => {
    if (player == Player.getName()) {
        utils.debugLog("You joined a party finder party")
        return
    }

    utils.debugLog("A player joined the party finder party")
    if (!isShitter(player)) return
    utils.debugLog("&cKicking player")
    reason = getShitterReason(player)
    ChatLib.command("pc [shaweelAddons] "+player+" is on my shitter list ("+reason+")")
    setTimeout(() => {
        ChatLib.command("party kick "+player)
    }, 1000)
}).setCriteria("Party Finder > ${player} joined the dungeon group! (${dungeonClass} Level ${level})")

//Get the reason for being a shitter
function getShitterReason(username) {
    username=username.toLowerCase()
    if (!isShitter(username)) return null
    for (let player of shitters) {
        if (player[0].toLowerCase() != username) continue
        return player[1]
    }
}

//Get if a player is on your shitterlist
function isShitter(username) {
    username = username.toLowerCase()
    for (let player of shitters) {
        if (player[0].toLowerCase() != username) continue
        return true
    }
    return false
}

//Save the shitterlist to the file
function saveShitters() {
    toSave = JSON.stringify(shitters)
    FileLib.write("./config/ChatTriggers/modules/shaweelAddons/shitters.json", toSave)
}

confirming = false
function shitter(arg2, arg3, reason) {
    //Prints the help message
    if (arg2 == undefined || arg2 == "help") {
        utils.chatLog("&c[Shitter List] &7Available commands:")
        ChatLib.chat("&a/sha shitterlist add <username> <reason(optional)>&7 - adds a player to your shitter list, with an optional reason.")
        ChatLib.chat("&a/sha shitterlist reason <username> <reason>&7 - adds/changes a player's reason for being on the shitter list.")
        ChatLib.chat("&a/sha shitterlist remove <username>&7 - removes a player from your shitter list.")
        ChatLib.chat("&a/sha shitterlist removeall&7 - removes all players from your shitter list, this action is irrevertable")
        ChatLib.chat("&a/sha shitterlist list&7 - lists every person on your shitter list, with the reason")
        ChatLib.chat("&a/sha shitterlist reload&7 - reloads the shitterlist using the "+"shitters.json"+"&7 file.")
        ChatLib.chat("&a/sha shitterlist help&7 - shows this help message.")
        ChatLib.chat("")
        ChatLib.chat("&7Every person on your shitter list will be auto kicked from your party whenever they join through party finder, when kicking it says that they got kicked and the reason for being on the shitter list in chat.")
        return
    }

    //Reloads the shitter list from the file
    if (arg2 == "reload") {
        shitters = FileLib.read("./config/ChatTriggers/modules/shaweelAddons/shitters.json")
        shitters = JSON.parse(shitters)
        utils.chatLog("&c[Shitter List] &7Succesfully reloaded the shitter list.")
        return
    }

    //Removes everyone from your shitterlist
    if (arg2 == "removeall") {
        confirming = true
        utils.chatLog("&c[Shitter List] &4&lTHIS ACTION IS IRREVERSIBLE&7, type &c/sha shitterlist confirm&7 in the next &a5&7 seconds to proceed.")
        setTimeout(() => {
            confirming = false
        }, 5000)
        return
    }

    //Confirmation for removing everyone from your shitterlist
    if (arg2 == "confirm" && confirming) {
        utils.chatLog("&c[Shitter List] &7Shitter list succesfully wiped.")
        shitters = []
        saveShitters()
        return
    }

    //Lists all players on your shitterlist
    if (arg2 == "list") {
        utils.chatLog("&c[Shitter List] &7All players on your shitter list:")
        for (let player of shitters) {
            ChatLib.chat("&e"+player[0]+"&7: "+player[1])
        }
        return
    }

    //Removes someone from your shitterlist
    if (arg2 == "remove") {
        if (arg3 == undefined) {
            utils.chatLog("&c[Shitter List] &7Usage: &a/sha shitterlist remove <player>")
            return
        }
        if (!isShitter(arg3)) {
            utils.chatLog("&c[Shitter List] &a"+arg3+"&7 is not on your shitter list.")
            return
        }
        utils.chatLog("&c[Shitter List] &7Succesfully removed &a"+arg3+"&7 from your shitter list.")
        n = null
        i=-1
        for (let player of shitters) {
            i+=1
            if (player[0].toLowerCase() != arg3.toLowerCase()) {
                utils.debugLog("&a"+player[0]+"&7 is not &a"+arg3)
                continue
            }
            utils.debugLog("Player to remove has been succesfully found.")
            n=i
            break
        }
        utils.debugLog("Attempting to remove &a"+arg3+" &7from the shitter list using the index &a"+n)
        shitters.splice(n, 1)
        saveShitters()
        return
    }

    //Changes the reason for someone being on your shitterlist
    if (arg2 == "reason") {
        if (arg3 == undefined || reason == "") {
            utils.chatLog("&c[Shitter List] &7Usage: &a/sha shitterlist reason <player> <reason>")
            return
        }

        if (!isShitter(arg3)) {
            utils.chatLog("&c[Shitter List] &a"+arg3+"&7 is not on your shitter list.")
            return
        }
        utils.chatLog("&c[Shitter List] &7Succesfully changed the reason of &a"+arg3+"&7 being on your shitter list to &a"+reason)
        for (let player of shitters) {
            if (player[0].toLowerCase() != arg3.toLowerCase()) {
                utils.debugLog("&a"+player[0]+"&7 is not &a"+arg3)
                continue
            }
            utils.debugLog("Player to change reason has been succesfully found.")
            player[1] = reason
            saveShitters()
            break
        }
        return
    } 

    //Adds someone to your shitterlist
    if (arg2 == "add") {
        if (arg3 == undefined) {
            utils.chatLog("&c[Shitter List] &7Usage: &a/sha shitterlist add <player> <reason(optional)>")
            return
        }

        if (reason == "") {
            reason = "No reason provided"
        }
        if (isShitter(arg3)) {
            utils.chatLog("&c[Shitter List] &a"+arg3+"&7 is already on your shitter list. Use &a/sha shitterlist reason&7, if you want to change the reason.")
            return
        }
        utils.chatLog("&c[Shitter List] &7Succesfully added &a"+arg3+"&7 to your shitter list because of: &a"+reason)
        shitters.push([arg3, reason])
        saveShitters()
        return
    }
    utils.chatLog("&c[Shitter List] &7Unknown command, use &a/sha shitterlist help&7 to learn more")
}

register("command", (...args) => {
    utils.debugLog("Argument 1: &a"+args[0])
    reason=args.slice(3).join(" ")
    index = -1

    //Prints the help message
    if (args[0] == "help") {
        utils.chatLog("Available commands:")
        ChatLib.chat("&a/sha config(or just /sha)&7 - Opens the config.")
        ChatLib.chat("&a/sha gui&7 - Opens the gui editor.")
        ChatLib.chat("&a/sha shitterlist(aliases: sl, shitter, slist, shitters)&7 - The local shitterlist, use &a/sha shitterlist&7 to learn more")
        ChatLib.chat("&a/sha debug&7 - Debug commands mainly used by shaweel to make the development process easier.")
        ChatLib.chat("&a/sha debugmode&7 - Gives extra information, that is useless if you're not working on the mod and just using it.")
        ChatLib.chat("&a/sha help&7 - Prints this help message.")
        return
    }

    //Toggles debug mode
    if (args[0] == "debugmode") {
        utils.toggleDebugMode()
        return
    }

    //Shitter list commands
    if (args[0] == "shitter" || args[0] == "shitterlist" || args[0] == "sl" || args[0] == "slist" || args[0] == "shitters") {
        shitter(args[1], args[2], reason)
        return
    }

    //Opens the GUI Editor
    if (args[0] == "gui") {
        startMovingGui()
        return
    }

    if (args[0] == "tyfr") {
        tyfr()
        return
    }

    //Opens the configuration
    if (args[0] == "config" || args[0] == undefined) {
        Settings.openGUI()
        return
    }
    utils.chatLog("&7Unknown command, use &a/sha help&7 to learn more")
}).setName("shaweeladdons").setAliases("shaweel", "sha", "shaddons", "saddons")


register("command", (...args) => {
    tyfr()
}).setName("tyfr")

File: src/main/kotlin/me/shaweel/shaweeladdons/commandsAndShitterlist.kt
package me.shaweel.shaweeladdons

import com.google.gson.JsonObject
import com.mojang.brigadier.Command
import com.mojang.brigadier.CommandDispatcher
import me.shaweel.shaweeladdons.utils.chatLog
import me.shaweel.shaweeladdons.utils.debugLog
import net.fabricmc.fabric.api.client.command.v2.ClientCommandRegistrationCallback
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource
import net.minecraft.client.MinecraftClient
import net.fabricmc.fabric.api.client.command.v2.ClientCommandManager
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.text.Text
import java.io.File
import java.nio.file.Files
import kotlin.io.path.Path
import kotlin.io.path.exists
import com.google.gson.JsonParser
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

object commandsAndShitterlist {
    val configFolder = FabricLoader.getInstance().configDir
    var shaConfigFolder = Path(configFolder.toString()+"/shaweelAddons")
    lateinit var debugModeFile: File
    lateinit var shitterListFile: File
    lateinit var shitterList: JsonObject
    var removing = false
    var debugMode = false
    val debugModeLiteral = ClientCommandManager.literal("debugmode").executes { context ->
        val player = MinecraftClient.getInstance().player
        debugMode = !debugMode
        debugModeFile.writeText(debugMode.toString())
        if (debugMode) {
            player?.sendMessage(Text.literal("§d[shaweelAddons]§e [DEBUG]§a Debug mode activated"), false)
        } else {
            player?.sendMessage(Text.literal("§d[shaweelAddons]§e [DEBUG]§c Debug mode deactivated"), false)
        }
        Command.SINGLE_SUCCESS
    }
    val gui = ClientCommandManager.literal("gui").executes { context ->
        debugLog("Opening the gui editor(I haven't made that yet)")
        openGuiEditor()
        Command.SINGLE_SUCCESS
    }
    val config = ClientCommandManager.literal("config").executes { context ->
        debugLog("Opening the ClickGui(I haven't made that yet)")
        openConfig()
        Command.SINGLE_SUCCESS
    }
    private fun help(alias: String): LiteralArgumentBuilder<FabricClientCommandSource?>? {
        return ClientCommandManager.literal("help").executes { context ->
            val helpMessage = """
                    Available commands:
                    §a/$alias config §7- Opens the config.
                    §a/$alias gui §7- Opens the gui editor.
                    §a/$alias shitterlist §7- The local shitterlist, do §a/$alias shitterlist help §7to learn more.
                    §a/$alias debugmode §7- Toggles debugmode which gives extra information, that is useless if you're not working on the mod and just using it.
                    §a/$alias help §7- Prints this help message.
                    """.trimIndent()
            chatLog(helpMessage)
            Command.SINGLE_SUCCESS
        }
    }
    private fun shitterList(alias: String, slAlias: String): LiteralArgumentBuilder<FabricClientCommandSource?>? {
        return ClientCommandManager.literal(slAlias).executes { context ->
            chatLog(getShitterHelpMessage(alias, slAlias))
            Command.SINGLE_SUCCESS
        }.then(ClientCommandManager.literal("help").executes {
            chatLog(getShitterHelpMessage(alias, slAlias))
            Command.SINGLE_SUCCESS
        }).then(ClientCommandManager.literal("list").executes {
            var msg = """§c[Shitter List] §7All players on your shitter list: """
            for ((name, reason) in shitterList.entrySet()) {
                msg+="\n§e$name§7: ${reason.asString}"
            }
            chatLog(msg)
            Command.SINGLE_SUCCESS
        }).then(ClientCommandManager.literal("removeall").executes {
            chatLog("§4§lTHIS ACTION IS IRREVERSIBLE§7, type §c/sha $slAlias confirm§7 in the next §a5 seconds to proceed.")
            removing = true
            GlobalScope.launch {
                delay(5000)
                removing = false
            }
            Command.SINGLE_SUCCESS
        }).then(ClientCommandManager.literal("confirm").executes {
            if (removing) {
                shitterList = JsonParser.parseString("{}").asJsonObject
                shitterListFile.writeText(shitterList.toString())
                chatLog("Successfully wiped the shitter list.")
            }
            Command.SINGLE_SUCCESS
        }).then(ClientCommandManager.literal("add")
            .then(ClientCommandManager.argument("username", StringArgumentType.string()).executes { context ->
                val username = StringArgumentType.getString(context, "username")
                val reason = "No reason provided."
                addToShitterlist(username, reason)
                Command.SINGLE_SUCCESS
            }.then(ClientCommandManager.argument("reason", StringArgumentType.greedyString()).executes { context ->
                val username = StringArgumentType.getString(context, "username")
                val reason = StringArgumentType.getString(context, "reason")
                addToShitterlist(username, reason)
                Command.SINGLE_SUCCESS
            })).executes {
                chatLog("§c[Shitter List] §7Usage: §a/$alias $slAlias add <username> <reason(optional)>")
                Command.SINGLE_SUCCESS
            }).then(ClientCommandManager.literal("reason")
            .then(ClientCommandManager.argument("username", StringArgumentType.string()).executes { context ->
                chatLog("§c[Shitter List] §7Usage: §a/$alias $slAlias reason <username> <reason>")
                Command.SINGLE_SUCCESS
            }.then(ClientCommandManager.argument("reason", StringArgumentType.greedyString()).executes { context ->
                val username = StringArgumentType.getString(context, "username")
                val reason = StringArgumentType.getString(context, "reason")
                if (!shitterList.entrySet().any { it.key.lowercase() == username.lowercase() }) {
                    chatLog("§c[Shitter List] §a$username §7is not on your shitter list.")
                    return@executes Command.SINGLE_SUCCESS
                }
                shitterList.addProperty(username, reason)
                shitterListFile.writeText(shitterList.toString())
                chatLog("§c[Shitter List] §7Successfully changed to reason of §a$username §7being on your shitter list to: §a$reason")
                Command.SINGLE_SUCCESS
            }))
            .executes {
                chatLog("§c[Shitter List] §7Usage: §a/$alias $slAlias reason <username> <reason>")
                Command.SINGLE_SUCCESS
            }
        ).then(ClientCommandManager.literal("remove")
            .then(ClientCommandManager.argument("username", StringArgumentType.string()).executes { context ->
                val username = StringArgumentType.getString(context, "username")
                val entry = shitterList.entrySet().find { it.key.lowercase() == username.lowercase() }
                if (entry == null) {
                    chatLog("§c[Shitter List] §a$username §7is not on your shitter list.")
                    return@executes Command.SINGLE_SUCCESS
                }
                shitterList.remove(entry.key)
                shitterListFile.writeText(shitterList.toString())
                chatLog("§c[Shitter List] §7Successfully removed §a$username §7from your shitter list.")
                Command.SINGLE_SUCCESS
            })
            .executes {
                chatLog("§c[Shitter List] §7Usage: §a/$alias $slAlias remove <username>")
                Command.SINGLE_SUCCESS
            }
        )
    }

    private fun openConfig() {}
    private fun openGuiEditor() {}

    private fun addToShitterlist(username: String, reason: String) {
        if (shitterList.entrySet().any { it.key.lowercase() == username.lowercase() }) {
            chatLog("§c[Shitter List] §a$username §7is already on your shitter list. Use §a/sha shitterlist reason§7, if you want to change the reason.")
            return
        }
        shitterList.addProperty(username, reason)
        shitterListFile.writeText(shitterList.toString())
        chatLog("§c[Shitter List] §7Successfully added §a$username §7to your shitter list because of: §a$reason")
    }

    private fun getShitterHelpMessage(alias: String, slAlias: String): String {
        return """
        §c[Shitter List] §7Available commands:
        §a/$alias $slAlias add <username> <reason(optional)>§7 - Adds a player to your shitter list, with an optional reason.
        §a/$alias $slAlias reason <username> <reason>§7 - Adds/changes a player's reason for being on the shitter list..
        §a/$alias $slAlias remove <username>§7 - Removes a player from your shitter list.
        §a/$alias $slAlias removeall§7 - Removes all players from your shitter list, this action is irreversible.
        §a/$alias $slAlias list§7 - Lists every person on your shitter list, with the reason.
        §a/$alias $slAlias help§7 - Prints this help message.
        
        §7Every person on your shitter list will be auto kicked from your party whenever they join through party finder, when kicking it says that they got kicked and the reason for being on the shitter list in chat.
        """.trimIndent()
    }

    private fun registerCommand(dispatcher: CommandDispatcher<FabricClientCommandSource>) {
        val aliases = listOf("shaweeladdons", "shaweel", "sha")
        for (alias in aliases) {
            val command = ClientCommandManager.literal(alias)
                .executes { context ->
                    debugLog("Opening the ClickGui(I haven't made that yet)")
                    openConfig()
                    Command.SINGLE_SUCCESS
                }
                .then(debugModeLiteral)
                .then(gui)
                .then(config)
                .then(help(alias))
                .then(shitterList(alias, "shitterlist"))
                .then(shitterList(alias, "sl"))
                .then(shitterList(alias, "shitter"))
            dispatcher.register(command)
        }
    }

    fun init() {
        ClientCommandRegistrationCallback.EVENT.register { dispatcher, _ ->
            registerCommand(dispatcher)
        }
        if (!shaConfigFolder.exists()) {
            shaConfigFolder = Files.createDirectory(Path(configFolder.toString()+"/shaweelAddons"))
        }
        debugModeFile = File(shaConfigFolder.toString()+"/debugmode.txt")
        if (!debugModeFile.exists()) {
            Files.createFile(Path(shaConfigFolder.toString()+"/debugmode.txt"))
            debugModeFile = File(shaConfigFolder.toString()+"/debugmode.txt")
            debugModeFile.writeText("false")
        }
        shitterListFile = File(shaConfigFolder.toString()+"/shitterlist.json")
        if (!shitterListFile.exists()) {
            Files.createFile(Path(shaConfigFolder.toString()+"/shitterlist.json"))
            shitterListFile = File(shaConfigFolder.toString()+"/shitterlist.json")
            shitterListFile.writeText("[]")
        }
        debugMode = debugModeFile.readText().toBoolean()
        val shitterListStr = shitterListFile.readText()
        shitterList = JsonParser.parseString(shitterListStr).asJsonObject
    }
}